rbx_binary/src/serializer.rs

```diff
@@ -803,6 +803,28 @@ impl<'a, W: Write> BinarySerializer<'a, W> {
 
                         chunk.write_interleaved_u32_array(&buf)?;
                     }
+                    Type::Ref => {
+                        let mut buf = Vec::with_capacity(values.len());
+                        let mut curr_ref;
+
+                        for (i, rbx_value) in values {
+                            if let Variant::Ref(value) = rbx_value.as_ref() {
+                                if value.is_none() {
+                                    curr_ref = -1;
+                                } else if let Some(id) = self.id_to_referent.get(value) {
+                                    curr_ref = *id
+                                } else {
+                                    return Err(InnerError::InvalidInstanceId { referent: *value });
+                                }
+
+                                buf.push(curr_ref);
+                            } else {
+                                return type_mismatch(i, &rbx_value, "Ref");
+                            }
+                        }
+
+                        chunk.write_referent_array(buf.into_iter())?;
```

Comment:
It doesn't look like we need to track `curr_ref` here anymore. Instead of creating a value, we should be able to just push directly into the buffer in both branches here.

Additionally, does referring to an invalid instance ID necessitate an error? If Roblox Studio saves a model that has a reference to an instance not included in the model, does a null referent get written, or does a referent get allocated for that value that is unused?

Reply:
Looks like Studio does write out the null referent in that case, so yeah, the third branch here can never be hit. I've removed it.